### PR TITLE
Add instantaneous, lagged and imaginary coherences

### PR DESCRIPTION
**Pull request by qbarthelemy**
The ordinary coherence is a normalization of cross-spectral matrices, capturing both in-phase and out-of-phase correlations. However it is inflated by the artificial in-phase (zero-lag) correlation engendered by volume conduction.

This PR adds:
- the instantaneous coherence capturing only in-phase correlation;
- the lagged-coherence capturing only out-of-phase correlation;
- the imaginary coherence capturing out-of-phase correlation but still affected by in-phase correlation.

### CODE REVIEW

File: pyriemann/estimation.py [hunk 1]
```diff
@@ -423,40 +422,82 @@ class Coherences(CospCovariances):
...
     CospCovariances
+
+    References
+    ----------
+    .. [1] Pascual-Marqui R. "Instantaneous and lagged measurements of linear
+        and nonlinear dependence between groups of multivariate time series:
+        frequency decomposition", arXiv, 2007
+
+    .. [2] Congedo, M. "Non-Parametric Synchronization Measures used in EEG
+        and MEG", TechReport, 2018
```
Comment (agramfort) [inline pyriemann/estimation.py:460]: to me https://pubmed.ncbi.nlm.nih.gov/15351371/ is another standard reference

File: tests/test_utils_covariance.py [hunk 2]
```diff
@@ -160,28 +160,39 @@ def test_covariances_cospectrum():
     assert_array_almost_equal(cosp_pr, cosp_sp, 6)
 
 
-def test_covariances_coherence():
+@pytest.mark.parametrize('coh', ['ordinary', 'instantaneous', 'lagged', 'imaginary', 'truc'])
```
Comment (agramfort) [inline tests/test_utils_covariance.py:163]:
```suggestion
@pytest.mark.parametrize('coh', ['ordinary', 'instantaneous', 'lagged', 'imaginary', 'foobar'])
```
more international...

File: tests/test_utils_covariance.py [hunk 3]
```diff
@@ -160,28 +160,39 @@ def test_covariances_cospectrum():
...
-        fs=fs,
-        nperseg=window,
-        noverlap=int(overlap * window),
-        window=np.hanning(window),
-        detrend=False)
-    # compare frequencies
-    assert_array_almost_equal(freqs_pr, freqs_sp, 6)
-    # compare coherence
-    assert_array_almost_equal(coh_pr[0, 1], coh_sp, 6)
+    if coh == 'truc':
```
Comment (agramfort) [inline tests/test_utils_covariance.py:170]:
```suggestion
    if coh == 'foobar':
```

File: pyriemann/estimation.py [hunk 4]
```diff
@@ -423,40 +422,82 @@ class Coherences(CospCovariances):
...
     --------
     Covariances
     HankelCovariances
     CospCovariances
+
+    References
+    ----------
+    .. [1] Pascual-Marqui R. "Instantaneous and lagged measurements of linear
+        and nonlinear dependence between groups of multivariate time series:
+        frequency decomposition", arXiv, 2007
```
Comment (agramfort) [inline pyriemann/estimation.py:457]: can you add the URL to pdf. Will make it easier to find equations

File: pyriemann/estimation.py (hunk 1 above)
Comment (agramfort) [reply pyriemann/estimation.py:460]: also can you add the URL to pdf. Will make it easier to find equations

File: tests/test_utils_covariance.py [hunk 5]
```diff
@@ -160,28 +160,39 @@ def test_covariances_cospectrum():
     assert_array_almost_equal(cosp_pr, cosp_sp, 6)
 
 
-def test_covariances_coherence():
+@pytest.mark.parametrize('coh', ['ordinary', 'instantaneous', 'lagged', 'imaginary', 'truc'])
+def test_covariances_coherence(coh):
```
Comment (agramfort) [inline tests/test_utils_covariance.py:164]: I would consider more ambitious tests eg that you actually test the statistical properties of the estimators eg that you have zero coh when you have a pure instantaneous mixing etc.

File: pyriemann/estimation.py (hunk 1 above)
Comment (qbarthelemy) [reply pyriemann/estimation.py:460]: As explained in Pascual Marqui 2007:

> While the instantaneous coherence is the real part of the complex valued coherency, the lagged coherence is not the imaginary part of the complex valued coherency. Ideally, the lagged coherence is a measure that *is not* affected by instantaneous dependence, whereas the imaginary part of the complex valued coherency (Nolte et al 2004) is more affected by instantaneous dependence (Pascual-Marqui 2007a).

So, with hindsight, imaginary coherence of Nolte seems an intermediary step towards the lagged-coherence. So, I am not sure that it is a standard reference.

Comment (agramfort) [reply pyriemann/estimation.py:460]: for me the Nolte paper is the seminal paper to have proposed the imaginary part of coherence. So at least I would point to Nolte et al for this option.

Comment (qbarthelemy) [reply pyriemann/estimation.py:460]: ok, fine.

File: tests/test_utils_covariance.py (hunk 5 above)
Comment (qbarthelemy) [reply tests/test_utils_covariance.py:166]: ok, but it requires time. wip

**Review by agramfort — APPROVED** [review]
thx @qbarthelemy 

merge if you're done.

**Comment by qbarthelemy** [comment]
I have completed tests.
@sylvchev , ok for you?
